numpy/core/src/multiarray/einsum.c.src

```diff
@@ -8,1917 +8,18 @@
-    &@name@_sum_of_products_contig_stride0_outstride0_two,
-    &@name@_sum_of_products_contig_stride0_outcontig_two,
-    &@name@_sum_of_products_contig_contig_outstride0_two,
-},
-#else
-    {NULL, NULL, NULL, NULL, NULL},
-#endif
-/**end repeat**/
-}; /* End of _binary_specialization_table */
-
-static sum_of_products_fn _outstride0_specialized_table[NPY_NTYPES][4] = {
-/**begin repeat
- * #name = bool,
- *         byte, ubyte,
- *         short, ushort,
- *         int, uint,
- *         long, ulong,
- *         longlong, ulonglong,
- *         float, double, longdouble,
- *         cfloat, cdouble, clongdouble,
- *         object, string, unicode, void,
- *         datetime, timedelta, half#
- * #use = 1,
- *        1, 1,
- *        1, 1,
- *        1, 1,
- *        1, 1,
- *        1, 1,
- *        1, 1, 1,
- *        1, 1, 1,
- *        0, 0, 0, 0,
- *        0, 0, 1#
- */
-#if @use@
-{
-    &@name@_sum_of_products_outstride0_any,
-    &@name@_sum_of_products_outstride0_one,
-    &@name@_sum_of_products_outstride0_two,
-    &@name@_sum_of_products_outstride0_three
-},
-#else
-    {NULL, NULL, NULL, NULL},
-#endif
-/**end repeat**/
-}; /* End of _outstride0_specialized_table */
-
-static sum_of_products_fn _allcontig_specialized_table[NPY_NTYPES][4] = {
-/**begin repeat
- * #name = bool,
- *         byte, ubyte,
- *         short, ushort,
- *         int, uint,
- *         long, ulong,
- *         longlong, ulonglong,
- *         float, double, longdouble,
- *         cfloat, cdouble, clongdouble,
- *         object, string, unicode, void,
- *         datetime, timedelta, half#
- * #use = 1,
- *        1, 1,
- *        1, 1,
- *        1, 1,
- *        1, 1,
- *        1, 1,
- *        1, 1, 1,
- *        1, 1, 1,
- *        0, 0, 0, 0,
- *        0, 0, 1#
- */
-#if @use@
-{
-    &@name@_sum_of_products_contig_any,
-    &@name@_sum_of_products_contig_one,
-    &@name@_sum_of_products_contig_two,
-    &@name@_sum_of_products_contig_three
-},
-#else
-    {NULL, NULL, NULL, NULL},
-#endif
-/**end repeat**/
-}; /* End of _allcontig_specialized_table */
-
-static sum_of_products_fn _unspecialized_table[NPY_NTYPES][4] = {
-/**begin repeat
- * #name = bool,
- *         byte, ubyte,
- *         short, ushort,
- *         int, uint,
- *         long, ulong,
- *         longlong, ulonglong,
- *         float, double, longdouble,
- *         cfloat, cdouble, clongdouble,
- *         object, string, unicode, void,
- *         datetime, timedelta, half#
- * #use = 1,
- *        1, 1,
- *        1, 1,
- *        1, 1,
- *        1, 1,
- *        1, 1,
- *        1, 1, 1,
- *        1, 1, 1,
- *        0, 0, 0, 0,
- *        0, 0, 1#
- */
-#if @use@
-{
-    &@name@_sum_of_products_any,
-    &@name@_sum_of_products_one,
-    &@name@_sum_of_products_two,
-    &@name@_sum_of_products_three
-},
-#else
-    {NULL, NULL, NULL, NULL},
-#endif
-/**end repeat**/
-}; /* End of _unnspecialized_table */
+#include "einsum_p.h"
```

Comment:
Can you make this change as a standalone PR? It makes it very difficult to review at the moment.

Reply:
(also, what does `_p` stand for?)

Reply:
> Can you make this change as a standalone PR? It makes it very difficult to review at the moment.

yes, it's possible to separate them into two parts:
 - 1. move to universal intrinsics
 - 2. runtime dispatching

Do you suggest it as a policy to follow or only within big patches?

> (also, what does _p stand for?)

private

Reply:
Sorry, what I'm suggesting here is make a new PR that simply moves all this existing code to the new file, without actually changing the implementation at all.

You can then add all the universal intrinsics / runtime dispatching in this PR, and github will actually show a useful diff.

Reply:
> > (also, what does _p stand for?)
> 
> private

I don't think we use that suffix anywhere else. Perhaps `einsum_helpers.h` or `einsum_loops.h` would be more appropriate?

Reply:
@eric-wieser,
> Sorry, what I'm suggesting here is make a new PR that simply moves all this existing code to the new file, without actually changing the implementation at all.

Swapping it, dispatching goes first then another pr for replacing the raw SIMD to universal intrinsics,
well, it's possible but I'm afraid that may add extra efforts to @Qiyu8.

> I don't think we use that suffix anywhere else. Perhaps einsum_helpers.h or einsum_loops.h would be more appropriate?

`_p` commonly used by `c/c++` community but `einsum_helpers.h` sounds good too

Reply:
> well, it's possible but I'm afraid that may add extra efforts to @Qiyu8.

That's why I'm suggesting the much simpler approach of:

1. Take all the code deleted in this file and put it in `einsum_helpers.h.src`, add the `#include`. Make a PR, commit message should indicate the code is moved without changes.
2. Rebase this PR using the `ours` strategy so that the contents of `einsum_helpers.h.src` is exactly as it is right now.

Step 1 doesn't need to be done by @Qiyu8, anyone could do it. Step 2 is easy, because @Qiyu8 at worst just pastes the current file content over whatever the git rebase gives them.

Reply:
@eric-wieser, sounds good to me, I assign my self to Step 1. @Qiyu8 Would you agree?

Reply:
@Qiyu8, I apologize for overriding you, however, please feel free to reject it if you have a different opinion.

@eric-wieser, I did some changes to the plan to fit the requirement of the dispatching process.
Now by assuming that #17107 is accepted,  @Qiyu8 gonna have to :
1. backup the content of `einsum.dispatch.c.src`
2. remove einsum.dispatch.c.src
3. remove `einsum_p.h.src`
4. reset `einsum.c.src` to the latest master
5. rebase to the latest master
6. replace the content of `einsum.dispatch.c.src` with the backup in the first step
7. push

Reply:
Let's wait for the merge of #17109.